corehq/apps/domain/forms.py

```diff
@@ -1157,3 +1167,430 @@ def process_submission(self, domain=None):
+                    'Please use this page only to extend the existing services contract.</p>'
+                    '</div>'
+                )),
+                self.form_actions
+            )
+
+    def process_subscription_management(self):
+        new_plan_version = DefaultProductPlan.get_default_plan_by_domain(
+            self.domain, edition=self.cleaned_data['software_plan_edition'],
+        )
+        revert_current_subscription_end_date = None
+        if self.current_subscription and self.cleaned_data['start_date'] < self.current_subscription.date_end:
+            revert_current_subscription_end_date = self.current_subscription.date_end
+            self.current_subscription.date_end = self.cleaned_data['start_date']
+            self.current_subscription.save()
+        try:
+            if not self.current_subscription or self.cleaned_data['start_date'] > datetime.date.today():
+                new_subscription = Subscription.new_domain_subscription(
+                    self.next_account,
+                    self.domain,
+                    new_plan_version,
+                    date_start=self.cleaned_data['start_date'],
+                    date_end=self.cleaned_data['end_date'],
+                    web_user=self.web_user,
+                )
+            else:
+                new_subscription = self.current_subscription.change_plan(
+                    new_plan_version,
+                    date_end=self.cleaned_data['end_date'],
+                    web_user=self.web_user,
+                )
+            if new_subscription.date_start <= datetime.date.today() and datetime.date.today() < new_subscription.date_end:
+                new_subscription.is_active = True
+            new_subscription.do_not_invoice = False
+            new_subscription.auto_generate_credits = True
+            new_subscription.save()
+        except:
+            # If the entire transaction did not go through, rollback saved changes
+            if revert_current_subscription_end_date:
+                self.current_subscription.date_end = revert_current_subscription_end_date
+                self.current_subscription.save()
+            raise
+
+        CreditLine.add_credit(
+            self.cleaned_data['sms_credits'],
+            feature_type=FeatureType.SMS,
+            subscription=new_subscription,
+            web_user=self.web_user,
+            reason=CreditAdjustmentReason.MANUAL,
+        )
+        CreditLine.add_credit(
+            self.cleaned_data['user_credits'],
+            feature_type=FeatureType.USER,
+            subscription=new_subscription,
+            web_user=self.web_user,
+            reason=CreditAdjustmentReason.MANUAL,
+        )
+
+    @property
+    def account_name(self):
+        return self.cleaned_data['fogbugz_client_name']
+
+    @property
+    def account_emails(self):
+        return self.cleaned_data['emails']
+
+    def clean_end_date(self):
+        end_date = self.cleaned_data['end_date']
```

Comment:
I know there was a bug before about `end_date` being None and then trying to compare `None` to a datetime. I saw that it wasn't a required field. let me know if this is something that could potentially happen